packages/babel-parser/src/plugins/estree.ts

```diff
@@ -175,6 +178,9 @@ export default (superClass: typeof Parser) =>
         end,
         afterBlockParse,
       );
+      if (node.directives.length) {
+        this.processComment(node);
```

Comment:
Currently `processComment` is only called in `finishNodeAt`. Here we call it before the `node` is finished so that `directiveToStmt` can manipulate the attached comments, _as if_ the `node` had been finished. And we are creating new AST nodes within `directiveToStmt`, I am worried that it will break the invariance assumed by the comment attaching algorithm elsewhere.

The reason of #14919 is that we are discarding the created AST nodes `Directive` and `DirectiveLiteral` and then creating new AST nodes to replace them. But the link from comment whitespaces to their leading nodes / trailing nodes are not updated, because at that point the tokenizer has moved _after_ the directives, and the comment processor is awaiting Program/Block to be finished so that the comment can be attached, assuming we are not manipulating the AST nodes ended before current position. So eventually the comment are attached to an unused AST node `Directive`, which is not in the final AST.

We can solve this issue by reusing the created AST node, specifically in
https://github.com/babel/babel/blob/4b7047dddcb09a93bc540cd48f461a8d503ce260/packages/babel-parser/src/plugins/estree.ts#L98

we can cast Directive to ExpressionStatement and DirectiveLiteral to Literal, update the AST shapes accordingly, essentially reverting what we did in the base parser:

https://github.com/babel/babel/blob/4b7047dddcb09a93bc540cd48f461a8d503ce260/packages/babel-parser/src/parser/statement.ts#L237

In this way the comment attaching should work for the ESTree directive representation.

Reply:
Many thanks for the detailed explanation, and for the pointer to `stmtToDirective()`! That saved me a ton of time.

Hopefully the revised code is in better shape.